File: test/Builtins/str/builtin_str.py
# RUN: test-ir.sh %s

print(str(1))
# IR: [[RV:%[0-9]+]] = call %PyObj* @llvmPy_str(%PyObj* %PyInt.1)
# IR: call %PyObj* @llvmPy_print(%PyObj* [[RV]])

# CHECK-DAG: declare %PyObj* @llvmPy_str(%PyObj*)

File: test/Builtins/str/builtin_str.cc
#include <llvmPy/Instr.h>
#include <llvmPy/PyObj/PyObj.h>
#include <catch2.h>
#include <fakeit.h>
using namespace llvmPy;
using namespace fakeit;

TEST_CASE("builtin: str()", "[builtins][str]") {
    Mock<PyObj> mock;
    std::string const str("test");

    SECTION("it delegates to the object's py__str__()") {
        When(Method(mock, py__str__)).Return(str);
        PyStr const *rv = llvmPy_str(mock.get());
        Verify(Method(mock, py__str__)).Once();
        VerifyNoOtherInvocations(mock);
        CHECK(rv->getValue() == str);
    }

}
